File: F-Droid/src/org/fdroid/fdroid/net/Downloader.java
package org.fdroid.fdroid.net;

import android.content.Context;
import android.content.Intent;
import android.support.v4.content.LocalBroadcastManager;
import android.util.Log;

import org.fdroid.fdroid.Utils;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.MalformedURLException;
import java.net.URL;

public abstract class Downloader {

    private static final String TAG = "Downloader";

    public static final String LOCAL_ACTION_PROGRESS = "Downloader.PROGRESS";

    public static final String EXTRA_ADDRESS = "extraAddress";
    public static final String EXTRA_BYTES_READ = "extraBytesRead";
    public static final String EXTRA_TOTAL_BYTES = "extraTotalBytes";

    private final OutputStream outputStream;

    private final LocalBroadcastManager localBroadcastManager;
    private final File outputFile;

    protected final URL sourceUrl;
    protected String cacheTag = null;
    protected int bytesRead = 0;
    protected int totalBytes = 0;

    public abstract InputStream getInputStream() throws IOException;

    Downloader(Context context, URL url, File destFile)
            throws FileNotFoundException, MalformedURLException {
        this.sourceUrl = url;
        outputFile = destFile;
        outputStream = new FileOutputStream(outputFile);
        localBroadcastManager = LocalBroadcastManager.getInstance(context);
    }

    /**
     * If you ask for the cacheTag before calling download(), you will get the
     * same one you passed in (if any). If you call it after download(), you
     * will get the new cacheTag from the server, or null if there was none.
     */
    public String getCacheTag() {
        return cacheTag;
    }

    /**
     * If this cacheTag matches that returned by the server, then no download will
     * take place, and a status code of 304 will be returned by download().
     */
    public void setCacheTag(String cacheTag) {
        this.cacheTag = cacheTag;
    }

    protected boolean wantToCheckCache() {
        return cacheTag != null;
    }

    /**
     * Only available if you passed a context object into the constructor
     * (rather than an outputStream, which may or  may not be associated with
     * a file).
     */
    public File getFile() {
        return outputFile;
    }

    public abstract boolean hasChanged();

    public abstract int totalDownloadSize();

    /**
     * Helper function for synchronous downloads (i.e. those *not* using AsyncDownloadWrapper),
     * which don't really want to bother dealing with an InterruptedException.
     * The InterruptedException thrown from download() is there to enable cancelling asynchronous
     * downloads, but regular synchronous downloads cannot be cancelled because download() will
     * block until completed.
     * @throws IOException
     */
    public void downloadUninterrupted() throws IOException {
        try {
            download();
        } catch (InterruptedException ignored) {}
    }

    public abstract void download() throws IOException, InterruptedException;

    public abstract boolean isCached();

    protected void downloadFromStream() throws IOException, InterruptedException {
        Utils.DebugLog(TAG, "Downloading from stream");
        InputStream input = null;
        try {
            input = getInputStream();

            // Getting the input stream is slow(ish) for HTTP downloads, so we'll check if
            // we were interrupted before proceeding to the download.
            throwExceptionIfInterrupted();

            copyInputToOutputStream(input);
        } finally {
            Utils.closeQuietly(outputStream);
            Utils.closeQuietly(input);
        }

        // Even if we have completely downloaded the file, we should probably respect
        // the wishes of the user who wanted to cancel us.
        throwExceptionIfInterrupted();
    }

    /**
     * In a synchronous download (the usual usage of the Downloader interface),
     * you will not be able to interrupt this because the thread will block
     * after you have called download(). However if you use the AsyncDownloadWrapper,
     * then it will use this mechanism to cancel the download.
     *
     * After every network operation that could take a while, we will check if an
     * interrupt occured during that blocking operation. The goal is to ensure we
     * don't move onto another slow, network operation if we have cancelled the
     * download.
     * @throws InterruptedException
     */
    private void throwExceptionIfInterrupted() throws InterruptedException {
        if (Thread.interrupted()) {
            Utils.DebugLog(TAG, "Received interrupt, cancelling download");
            throw new InterruptedException();
        }
    }

    /**
     * This copies the downloaded data from the InputStream to the OutputStream,
     * keeping track of the number of bytes that have flowed through for the
     * progress counter.
     */
    protected void copyInputToOutputStream(InputStream input) throws IOException, InterruptedException {

        byte[] buffer = new byte[Utils.BUFFER_SIZE];
        int bytesRead = 0;
        this.totalBytes = totalDownloadSize();

        // Getting the total download size could potentially take time, depending on how
        // it is implemented, so we may as well check this before we proceed.
        throwExceptionIfInterrupted();

        sendProgress(bytesRead, totalBytes);
        while (true) {

            int count = input.read(buffer);
            throwExceptionIfInterrupted();

            if (count == -1) {
                Utils.DebugLog(TAG, "Finished downloading from stream");
                break;
            }

            bytesRead += count;
            sendProgress(bytesRead, totalBytes);
            outputStream.write(buffer, 0, count);
        }
        outputStream.flush();
    }

    protected void sendProgress(int bytesRead, int totalBytes) {
        this.bytesRead = bytesRead;
        Intent intent = new Intent(LOCAL_ACTION_PROGRESS);
        intent.putExtra(EXTRA_ADDRESS, sourceUrl.toString());
        intent.putExtra(EXTRA_BYTES_READ, bytesRead);
        intent.putExtra(EXTRA_TOTAL_BYTES, totalBytes);
        localBroadcastManager.sendBroadcast(intent);
    }

    public int getBytesRead() {
        return bytesRead;
    }

    public int getTotalBytes() {
        return totalBytes;
    }
}

File: libs/download/src/androidMain/kotlin/org/fdroid/download/Downloader.kt
package org.fdroid.download

import mu.KotlinLogging
import org.fdroid.IndexFile
import org.fdroid.fdroid.ProgressListener
import org.fdroid.fdroid.isMatching
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException
import java.io.InputStream
import java.io.OutputStream
import java.security.MessageDigest

public abstract class Downloader(
    protected val indexFile: IndexFile,
    @JvmField
    protected val outputFile: File,
) {

    public companion object {
        private val log = KotlinLogging.logger {}
    }

    /**
     * If you ask for the cacheTag before calling download(), you will get the
     * same one you passed in (if any). If you call it after download(), you
     * will get the new cacheTag from the server, or null if there was none.
     *
     * If this cacheTag matches that returned by the server, then no download will
     * take place, and a status code of 304 will be returned by download().
     */
    @Deprecated("Used only for v1 repos")
    public var cacheTag: String? = null

    @Volatile
    private var cancelled = false

    @Volatile
    private var progressListener: ProgressListener? = null

    /**
     * Call this to start the download.
     * Never call this more than once. Create a new [Downloader], if you need to download again!
     */
    @Throws(IOException::class, InterruptedException::class, NotFoundException::class)
    public abstract fun download()

    @Throws(IOException::class, NotFoundException::class)
    protected abstract fun getInputStream(resumable: Boolean): InputStream
    protected open suspend fun getBytes(resumable: Boolean, receiver: BytesReceiver) {
        throw NotImplementedError()
    }

    /**
     * Returns the size of the file to be downloaded in bytes.
     * Note this is -1 when the size is unknown.
     * Used only for progress reporting.
     */
    protected abstract fun totalDownloadSize(): Long

    /**
     * After calling [download], this returns true if a new file was downloaded and
     * false if the file on the server has not changed and thus was not downloaded.
     */
    @Deprecated("Only for v1 repos")
    public abstract fun hasChanged(): Boolean
    public abstract fun close()

    public fun setListener(listener: ProgressListener) {
        progressListener = listener
    }

    @Throws(IOException::class, InterruptedException::class)
    protected fun downloadFromStream(isResume: Boolean) {
        log.debug { "Downloading from stream" }
        try {
            FileOutputStream(outputFile, isResume).use { outputStream ->
                getInputStream(isResume).use { input ->
                    // Getting the input stream is slow(ish) for HTTP downloads, so we'll check if
                    // we were interrupted before proceeding to the download.
                    throwExceptionIfInterrupted()
                    copyInputToOutputStream(input, outputStream)
                }
            }
            // Even if we have completely downloaded the file, we should probably respect
            // the wishes of the user who wanted to cancel us.
            throwExceptionIfInterrupted()
        } finally {
            close()
        }
    }

    @Suppress("BlockingMethodInNonBlockingContext")
    @Throws(
        InterruptedException::class,
        IOException::class,
        NoResumeException::class,
        NotFoundException::class,
    )
    protected suspend fun downloadFromBytesReceiver(isResume: Boolean) {
        try {
            val messageDigest: MessageDigest? = if (indexFile.sha256 == null) null else {
                MessageDigest.getInstance("SHA-256")
            }
            var bytesCopied = outputFile.length()
            // read pre-downloaded bytes (if any) for hash to match
            if (bytesCopied > 0 && messageDigest != null) outputFile.initDigest(messageDigest)
            FileOutputStream(outputFile, isResume).use { outputStream ->
                var lastTimeReported = 0L
                val bytesTotal = totalDownloadSize()
                getBytes(isResume) { bytes, numTotalBytes ->
                    // Getting the input stream is slow(ish) for HTTP downloads, so we'll check if
                    // we were interrupted before proceeding to the download.
                    throwExceptionIfInterrupted()
                    outputStream.write(bytes)
                    messageDigest?.update(bytes)
                    bytesCopied += bytes.size
                    val total = if (bytesTotal == -1L) numTotalBytes ?: -1L else bytesTotal
                    lastTimeReported = reportProgress(lastTimeReported, bytesCopied, total)
                }
                // check if expected sha256 hash matches
                indexFile.sha256?.let { expectedHash ->
                    if (!messageDigest.isMatching(expectedHash)) {
                        throw IOException("Hash not matching")
                    }
                }
                // force progress reporting at the end
                reportProgress(0L, bytesCopied, bytesTotal)
            }
            // Even if we have completely downloaded the file, we should probably respect
            // the wishes of the user who wanted to cancel us.
            throwExceptionIfInterrupted()
        } finally {
            close()
        }
    }

    /**
     * This copies the downloaded data from the [InputStream] to the [OutputStream],
     * keeping track of the number of bytes that have flown through for the [progressListener].
     *
     * Attention: The caller is responsible for closing the streams.
     */
    @Throws(IOException::class, InterruptedException::class)
    private fun copyInputToOutputStream(input: InputStream, output: OutputStream) {
        val messageDigest: MessageDigest? = if (indexFile.sha256 == null) null else {
            MessageDigest.getInstance("SHA-256")
        }
        try {
            var bytesCopied = outputFile.length()
            // read pre-downloaded bytes (if any) for hash to match
            if (bytesCopied > 0 && messageDigest != null) outputFile.initDigest(messageDigest)

            var lastTimeReported = 0L
            val bytesTotal = totalDownloadSize()
            val buffer = ByteArray(DEFAULT_BUFFER_SIZE)
            var numBytes = input.read(buffer)
            while (numBytes >= 0) {
                throwExceptionIfInterrupted()
                output.write(buffer, 0, numBytes)
                messageDigest?.update(buffer, 0, numBytes)
                bytesCopied += numBytes
                lastTimeReported = reportProgress(lastTimeReported, bytesCopied, bytesTotal)
                numBytes = input.read(buffer)
            }
            // check if expected sha256 hash matches
            indexFile.sha256?.let { expectedHash ->
                if (!messageDigest.isMatching(expectedHash)) {
                    throw IOException("Hash not matching")
                }
            }
            // force progress reporting at the end
            reportProgress(0L, bytesCopied, bytesTotal)
        } finally {
            output.flush()
            progressListener = null
        }
    }

    private fun reportProgress(lastTimeReported: Long, bytesRead: Long, bytesTotal: Long): Long {
        val now = System.currentTimeMillis()
        return if (now - lastTimeReported > 1000) {
            log.debug { "onProgress: $bytesRead/$bytesTotal" }
            progressListener?.onProgress(bytesRead, bytesTotal)
            now
        } else {
            lastTimeReported
        }
    }

    /**
     * Cancel a running download, triggering an [InterruptedException]
     */
    public fun cancelDownload() {
        cancelled = true
    }

    /**
     * Check if the download was cancelled.
     */
    public fun wasCancelled(): Boolean {
        return cancelled
    }

    /**
     * After every network operation that could take a while, we will check if an
     * interrupt occurred during that blocking operation. The goal is to ensure we
     * don't move onto another slow, network operation if we have cancelled the
     * download.
     *
     * @throws InterruptedException
     */
    @Throws(InterruptedException::class)
    private fun throwExceptionIfInterrupted() {
        if (cancelled) {
            log.info { "Received interrupt, cancelling download" }
            Thread.currentThread().interrupt()
            throw InterruptedException()
        }
    }

    @Throws(IOException::class)
    private fun File.initDigest(messageDigest: MessageDigest) {
        FileInputStream(this).use { inputStream ->
            val buffer = ByteArray(DEFAULT_BUFFER_SIZE)
            var bytes = inputStream.read(buffer)
            while (bytes >= 0) {
                messageDigest.update(buffer, 0, bytes)
                bytes = inputStream.read(buffer)
            }
        }
    }

}
